File: java/src/main/java/com/workshop/architecture/fitness/handle_payment_received/HandlePaymentReceivedController.java
package com.workshop.architecture.fitness.handle_payment_received;

import com.workshop.architecture.fitness.shared.MembershipBillingReferenceEntity;
import com.workshop.architecture.fitness.shared.MembershipBillingReferenceRepository;
import com.workshop.architecture.fitness.shared.MembershipEntity;
import com.workshop.architecture.fitness.shared.MembershipRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

import java.time.Instant;
import java.time.ZoneOffset;
import java.util.UUID;

@RestController
@RequestMapping("/api/memberships")
public class HandlePaymentReceivedController {

    private final MembershipRepository membershipRepository;
    private final MembershipBillingReferenceRepository billingReferenceRepository;

    public HandlePaymentReceivedController(
            MembershipRepository membershipRepository,
            MembershipBillingReferenceRepository billingReferenceRepository
    ) {
        this.membershipRepository = membershipRepository;
        this.billingReferenceRepository = billingReferenceRepository;
    }


    @PostMapping("/payment-received")
    ResponseEntity<PaymentReceivedResponse> paymentReceived(@RequestBody PaymentReceivedRequest request) {
        MembershipBillingReferenceEntity billingReference;
        MembershipEntity membership;
        Instant paidAt;
        String previousMembershipStatus;
        String newMembershipStatus;
        String message;
        boolean reactivated;

        if ((request.externalInvoiceId() == null || request.externalInvoiceId().isBlank())
                && (request.externalInvoiceReference() == null || request.externalInvoiceReference().isBlank())
                && (request.membershipId() == null || request.membershipId().isBlank())) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_REQUEST,
                    "At least one invoice or membership identifier must be provided"
            );
        }

        if (request.externalInvoiceId() != null && !request.externalInvoiceId().isBlank()) {
            billingReference = billingReferenceRepository.findByExternalInvoiceId(request.externalInvoiceId())
                    .orElse(null);
        } else {
            billingReference = null;
        }

        if (billingReference == null
                && request.externalInvoiceReference() != null
                && !request.externalInvoiceReference().isBlank()) {
            billingReference = billingReferenceRepository.findByExternalInvoiceReference(
                    request.externalInvoiceReference()
            ).orElse(null);
        }

        if (billingReference == null && request.membershipId() != null && !request.membershipId().isBlank()) {
            billingReference = billingReferenceRepository.findByMembershipId(UUID.fromString(request.membershipId()))
                    .stream()
                    .findFirst()
                    .orElse(null);
        }

        if (billingReference == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "No billing reference was found");
        }

        paidAt = request.paidAt() == null ? Instant.now() : request.paidAt();
        message = billingReference.isPaid()
                ? "Payment was already recorded; membership status unchanged"
                : "Payment recorded; membership status unchanged";

        if (!billingReference.isPaid()) {
            billingReference.markPaid(paidAt);
            billingReference = billingReferenceRepository.save(billingReference);
        }

        UUID billingReferenceMembershipId = billingReference.getMembershipId();

        membership = membershipRepository.findById(billingReferenceMembershipId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Membership %s was not found".formatted(billingReferenceMembershipId)
                ));

        previousMembershipStatus = membership.getStatus();
        newMembershipStatus = membership.getStatus();
        reactivated = false;

        if (membership.isCancelled()) {
            message = "Payment recorded; membership is cancelled and remains unchanged";
        } else if (membership.isSuspendedForNonPayment()) {
            if (!paidAt.atZone(ZoneOffset.UTC).toLocalDate().isAfter(membership.getEndDate())) {
                membership.reactivateAfterPayment();
                membership = membershipRepository.save(membership);
                newMembershipStatus = membership.getStatus();
                message = "Payment recorded; membership reactivated";
                reactivated = true;
            }
        }

        return ResponseEntity.ok(new PaymentReceivedResponse(
                paidAt,
                membership.getId().toString(),
                billingReference.getId().toString(),
                previousMembershipStatus,
                newMembershipStatus,
                reactivated,
                message
        ));
    }
}

File: python/src/workshop_api/fitness/managing_memberships/handle_payment_received/handle_payment_received_controller.py
from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from workshop_api.fitness.database import get_db_session

from ..shared.membership_entity import (
    MembershipBillingReferenceOrmModel,
    MembershipOrmModel,
)
from .payment_received_request import (
    PaymentReceivedRequest,
)
from .payment_received_response import (
    PaymentReceivedResponse,
)

router = APIRouter(prefix="/api/memberships", tags=["membership"])


@router.post(
    "/payment-received",
    response_model=PaymentReceivedResponse,
    response_model_by_alias=True,
)
def payment_received(
    request: PaymentReceivedRequest,
    session: Session = Depends(get_db_session),
) -> PaymentReceivedResponse:
    if (
        _is_blank(request.external_invoice_id)
        and _is_blank(request.external_invoice_reference)
        and _is_blank(request.membership_id)
    ):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="At least one invoice or membership identifier must be provided",
        )

    billing_reference = None
    if not _is_blank(request.external_invoice_id):
        billing_reference = (
            session.query(MembershipBillingReferenceOrmModel)
            .filter(
                MembershipBillingReferenceOrmModel.external_invoice_id
                == request.external_invoice_id
            )
            .first()
        )

    if billing_reference is None and not _is_blank(request.external_invoice_reference):
        billing_reference = (
            session.query(MembershipBillingReferenceOrmModel)
            .filter(
                MembershipBillingReferenceOrmModel.external_invoice_reference
                == request.external_invoice_reference
            )
            .first()
        )

    if billing_reference is None and not _is_blank(request.membership_id):
        billing_reference = (
            session.query(MembershipBillingReferenceOrmModel)
            .filter(MembershipBillingReferenceOrmModel.membership_id == request.membership_id)
            .first()
        )

    if billing_reference is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No billing reference was found",
        )

    paid_at = request.paid_at if request.paid_at is not None else datetime.now(UTC)
    message = (
        "Payment was already recorded; membership status unchanged"
        if billing_reference.is_paid()
        else "Payment recorded; membership status unchanged"
    )

    if not billing_reference.is_paid():
        billing_reference.mark_paid(paid_at)
        session.commit()
        session.refresh(billing_reference)

    membership = session.get(MembershipOrmModel, billing_reference.membership_id)
    if membership is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Membership {billing_reference.membership_id} was not found",
        )

    previous_membership_status = membership.status
    new_membership_status = membership.status
    reactivated = False

    if membership.is_cancelled():
        message = "Payment recorded; membership is cancelled and remains unchanged"
    elif membership.is_suspended_for_non_payment() and _as_utc_date(paid_at) <= membership.end_date:
        membership.reactivate_after_payment()
        session.commit()
        session.refresh(membership)
        new_membership_status = membership.status
        message = "Payment recorded; membership reactivated"
        reactivated = True

    return PaymentReceivedResponse(
        paidAt=paid_at,
        membershipId=membership.id,
        billingReferenceId=billing_reference.id,
        previousMembershipStatus=previous_membership_status,
        newMembershipStatus=new_membership_status,
        reactivated=reactivated,
        message=message,
    )


def _is_blank(value: str | None) -> bool:
    return value is None or value.strip() == ""


def _as_utc_date(value: datetime):
    if value.tzinfo is None:
        return value.date()
    return value.astimezone(UTC).date()
